confluent_kafka/schema_registry/protobuf.py

```diff
@@ -0,0 +1,437 @@
+                buf.write(_bytes((value & 0x7f) | 0x80))
+                value >>= 7
+            buf.write(_bytes(value))
+
+    def resolve_dependencies(self, ctx, file_desc):
+        """
+        Resolves and optionally registers schema references recursively.
+
+        Args:
+            ctx (SerializationContext): Serialization context.
+
+            file_desc (FileDescriptor): file descriptor to traverse.
+
+        """
+        schema_refs = []
+        for dep in file_desc.dependencies:
+            dep_refs = self.resolve_dependencies(ctx, dep)
+            subject = self._ref_reference_subject_func(ctx, dep)
+            schema = Schema(_schema_to_str(dep),
+                            references=dep_refs,
+                            schema_type='PROTOBUF')
+            if self._auto_register:
+                self._registry.register_schema(subject, schema)
+
+            reference = self._registry.lookup_schema(subject, schema)
+            # schema_refs are per file descriptor
+            schema_refs.append(SchemaReference(dep.name,
+                                               subject,
+                                               reference.version))
+        return schema_refs
+
+    def __call__(self, msg, ctx):
+        """
+        Serializes a Protobuf Message to the Confluent Schema Registry
+        Protobuf binary format.
+
+        Args:
+            msg (Message): Protobuf message instance.
+
+            ctx (SerializationContext): Metadata pertaining to the serialization
+                operation.
+
+        Note:
+            None objects are represented as Kafka Null.
+
+        Raises:
+            SerializerError if any error occurs serializing obj
+
+        Returns:
+            bytes: Confluent Schema Registry formatted Protobuf bytes
+
+        """
+        if msg is None:
+            return None
+
+        subject = self._subject_name_func(ctx, msg.DESCRIPTOR.full_name)
+
+        if subject not in self._known_subjects:
+            self._schema.references = self.resolve_dependencies(ctx, msg.DESCRIPTOR.file)
+
+        self._schema_id = self._registry.register_schema(subject, self._schema)
+
+        with _ContextStringIO() as fo:
+            # Write the magic byte and schema ID in network byte order (big endian)
+            fo.write(struct.pack('>bI', _MAGIC_BYTE, self._schema_id))
+            # write the record index to the buffer
+            self._encode_index(fo, self.msg_index)
+            # write the record itself
+            fo.write(msg.SerializeToString())
+            return fo.getvalue()
+
+
+class ProtobufDeserializer(object):
+    """
+    ProtobufDeserializer decodes bytes written in the Schema Registry
+    Protobuf format to an object.
+
+    Args:
+        descriptor (Descriptor): Protobuf Message Descriptor.
+
+    .. _Schema definition:
+        https://json-schema.org/understanding-json-schema/reference/generic.html
+
+    """
+    def __init__(self, descriptor):
```

Comment:
i think we should not pass descriptors in here.

Reply:
Related to first comment, reached out in slack to spark up conversation about more flexible protobuf serde. The 1:1 instance type requirement brought this about.

Reply:
We will revisit with the follow up PR which moves away from binding the  serializer to a specific protobuf type.